src/Microsoft.Framework.DesignTimeHost/ProtocolManager.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Framework.DesignTimeHost.Models;
+using Microsoft.Framework.Runtime;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Framework.DesignTimeHost
+{
+    public class ProtocolManager
+    {
+        /// <summary>
+        /// Environment variable for overriding protocol.
+        /// </summary>
+        public const string EnvDthProtocol = "DTH_PROTOCOL";
+
+        /// <summary>
+        /// Type name of the message representing the protocol negotiation.
+        /// </summary>
+        public const string NegotiationMessageTypeName = "ProtocolVersion";
+
+        public ProtocolManager(int capVersion)
+        {
+            CapVersion = capVersion;
+
+            // initialized to the highest supported version or environment overridden value
+            int? protocol = GetProtocolVersionFromEnvironment();
+
+            if (protocol.HasValue)
+            {
+                CurrentVersion = protocol.Value;
+                EnvironmentOverridden = true;
+            }
+            else
+            {
+                CurrentVersion = capVersion;
+            }
+        }
+
+        public int CapVersion { get; }
```

Comment:
Yes. The highest allowed version is decided when the `ProtocolManager` is created. The consumer holds the reference to `ProtocolManager` to reference the version.